File: apps/minecraft/spigot/src/main/java/crabcraft/net/crabUtilities/voicechat/VoicechatIntegration.java
package crabcraft.net.crabUtilities.voicechat;

import crabcraft.net.crabUtilities.CrabUtilities;
import de.maxhenkel.voicechat.api.BukkitVoicechatService;
import org.bukkit.event.EventPriority;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerQuitEvent;

/** Loads Simple Voice Chat API types only after the soft dependency is present. */
public final class VoicechatIntegration {

    private VoicechatIntegration() {}

    public static AutoCloseable register(CrabUtilities plugin) {
        BukkitVoicechatService service = plugin.getServer()
                .getServicesManager()
                .load(BukkitVoicechatService.class);
        if (service == null) {
            return null;
        }

        CrabVoicechatPlugin voicechatPlugin = new CrabVoicechatPlugin(plugin);
        service.registerPlugin(voicechatPlugin);
        Listener quitListener = new Listener() {};
        plugin.getServer().getPluginManager().registerEvent(
                PlayerQuitEvent.class, quitListener, EventPriority.LOWEST,
                (listener, event) -> voicechatPlugin.beforePlayerQuit(
                        ((PlayerQuitEvent) event).getPlayer().getUniqueId()),
                plugin);
        plugin.getLogger().info("Registered Simple Voice Chat plugin");

        return () -> {
            try {
                voicechatPlugin.shutdown();
            } finally {
                HandlerList.unregisterAll(quitListener);
                plugin.getServer().getServicesManager().unregister(voicechatPlugin);
            }
        };
    }
}

File: apps/minecraft/spigot/src/main/kotlin/crabcraft/net/crabUtilities/voicechat/VoicechatIntegration.kt
package crabcraft.net.crabUtilities.voicechat

import crabcraft.net.crabUtilities.CrabUtilities
import de.maxhenkel.voicechat.api.BukkitVoicechatService
import org.bukkit.event.EventPriority
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerQuitEvent

/** Loads Simple Voice Chat API types only after the soft dependency is present. */
object VoicechatIntegration {
    @JvmStatic
    fun register(plugin: CrabUtilities): AutoCloseable? {
        val service = plugin.server.servicesManager.load(BukkitVoicechatService::class.java) ?: return null
        val voicechatPlugin = CrabVoicechatPlugin(plugin)
        service.registerPlugin(voicechatPlugin)
        val quitListener = object : Listener {}
        plugin.server.pluginManager.registerEvent(
            PlayerQuitEvent::class.java,
            quitListener,
            EventPriority.LOWEST,
            { _, event -> voicechatPlugin.beforePlayerQuit((event as PlayerQuitEvent).player.uniqueId) },
            plugin,
        )
        plugin.logger.info("Registered Simple Voice Chat plugin")
        return AutoCloseable {
            try {
                voicechatPlugin.shutdown()
            } finally {
                HandlerList.unregisterAll(quitListener)
                plugin.server.servicesManager.unregister(voicechatPlugin)
            }
        }
    }
}
